vllm/model_executor/models/eagle.py

```diff
@@ -0,0 +1,140 @@
+        # -up.
+        self.token_map = None
+
+    @property
+    def sampler(self):
+        return self.model.sampler
+
+    def forward(
+        self,
+        input_ids: torch.Tensor,
+        positions: torch.Tensor,
+        kv_caches: List[torch.Tensor],
+        attn_metadata: AttentionMetadata,
+        previous_hidden_states: torch.Tensor,
+        intermediate_tensors: Optional[IntermediateTensors] = None,
+    ) -> torch.Tensor:
+
+        tok_embeds = self.model.model.embed_tokens(input_ids)
+        inputs_embeds = self.fc(
+            torch.cat([tok_embeds, previous_hidden_states], dim=-1))
+
+        inputs_embeds[positions == 0] = 0  # masking inputs at position=0
+
+        hidden_states = self.model.model(
+            input_ids=None,
+            inputs_embeds=inputs_embeds,
+            positions=positions,
+            kv_caches=kv_caches,
+            attn_metadata=attn_metadata,
+            intermediate_tensors=intermediate_tensors)
+        return hidden_states
+
+    def compute_logits(self, hidden_states: torch.Tensor,
+                       sampling_metadata: SamplingMetadata) -> torch.Tensor:
+        logits = self.logits_processor(self.lm_head, hidden_states,
+                                       sampling_metadata)
+
+        if self.token_map is not None:
+            _logits = logits
+            logits = -torch.inf * torch.ones(
+                size=(*_logits.shape[:-1], self.orig_vocab_size),
+                device=_logits.device,
+                dtype=_logits.dtype)
+
+            logits[..., self.token_map] = _logits
+
+        return logits
+
+    def sample(
+        self,
+        logits: torch.Tensor,
+        sampling_metadata: SamplingMetadata,
+    ) -> Optional[SamplerOutput]:
+        next_tokens = self.sampler(logits, sampling_metadata)
+        return next_tokens
+
+    def load_weights(self, weights: Iterable[Tuple[str, torch.Tensor]]):
+        model_weights = {}
+        for name, loaded_weight in weights:
+            if name == "token_map":
+                if self.config.truncated_vocab_size < self.config.vocab_size:
+                    self.token_map = nn.Parameter(loaded_weight,
+                                                  requires_grad=False)
+            elif name.startswith("fc."):
+                weight_loader = getattr(self.fc.weight, "weight_loader",
+                                        default_weight_loader)
+                weight_loader(self.fc.weight, loaded_weight)
+            elif name.startswith("model.lm_head.") or name.startswith(
+                    "model.model."):
+                model_weights[name.split("model.", 1)[-1]] = loaded_weight
+            elif name.startswith("lm_head.") or name.startswith("model."):
+                model_weights[name] = loaded_weight
+            else:
+                model_weights[f"model.{name}"] = loaded_weight
+
+        lm_head_weight = model_weights.pop("lm_head.weight")
```

Comment:
I'm testing with eagle llama3, but it seems there are still compatibility issues.
The `EAGLE-LLaMA3-Instruct-8B` model has an error because there is no lm_head in that part.
Could you share the test results for the llama3 8b model?

Reply:
Yes, it's because this model doesn't have lm_head. But since this was trained using the frozen lm_head from target model, we can add it manually from Llama3 checkpoint. I am uploading a version here: https://huggingface.co/abhigoyal/EAGLE-LLaMA3-Instruct-8B-vllm. Please try that. As I mentioned in another comment, this still doesn't give speed-up yet, but we have seen 1.3x seed-up in internal experiments.

Reply:
I was able to get the e2e tests (tests/spec_decode/e2e/test_eagle_correctness.py) to work with MAIN_MODEL = "meta-llama/Meta-Llama-3-8B" and SPEC_MODEL = "abhigoyal/EAGLE-LLaMA3-Instruct-8B-vllm". However I am wondering if it makes sense to add a comment regarding the incompatibility of yuhuili/EAGLE-LLaMA3-Instruct-8B with this implementation. Also if possible can you add a pointer to the example implementation of how you generated the working checkpoint (abhigoyal/EAGLE-LLaMA3-Instruct-8B-vllm) from the original one.

Reply:
Sure, I'll add the comments and explanation on how to create the working checkpoint.

Reply:
![plot_ttft_decode_thru](https://github.com/user-attachments/assets/84aa0e85-1be0-438e-889a-8d7b536c2a30)
The above results are graphs showing Time To First Token (TTFT) and Decode Throughput.
I conducted a performance comparison experiment with a fixed number of concurrent users, running a total of five configurations(one without speculative decoding and four with `--num-speculative-tokens` set to 2, 3, 4, and 5).

Using a single A100 GPU with LLaMA3 8B, I observed that using `--num-speculative-tokens 5` resulted in a lower draft acceptance rate, making it difficult to achieve significant speed improvements.

Reply:
@jacob-crux Thanks for trying it out and sharing these results. These are very helpful. May I ask what dataset did you use for this?

Reply:
@abhigoyal1997 
I used sharegpt dataset.

Reply:
How does the draft acceptance rate look? A while ago I checked top1 acceptance rate for Eagle/llama2 and it was about .7 on sharegpt. This is considerably better than a 68m draft model which got .5-.6, and Medusa's top5 of .8.

Reply:
Ran the serving_tests in nightly-benchmarks with the config below and found the following acceptance rates. I used the target model as "meta-llama/Meta-Llama-3-8B-Instruct" and draft model as "abhigoyal/EAGLE-LLaMA3-Instruct-8B-vllm"

k=1 Draft acceptance rate: 0.637, System efficiency: 0.819

k=2 Draft acceptance rate: 0.445, System efficiency: 0.594

k=3  Draft acceptance rate: 0.377, System efficiency: 0.463

k=4  Draft acceptance rate: 0.335, System efficiency: 0.374

The acceptance rate is lower than what you observed especially at larger k.

```
[
    {
        "test_name": "serving_llama8B_tp4_sharegpt_specdecode",
        "qps_list": [1],
        "server_parameters": {
            "model": "meta-llama/Meta-Llama-3-8B-Instruct",
            "disable_log_requests": "", 
            "tensor_parallel_size": 4,
            "swap_space": 16, 
            "speculative_model": "abhigoyal/EAGLE-LLaMA3-Instruct-8B-vllm",
            "num_speculative_tokens": 3,
            "speculative_draft_tensor_parallel_size": 1,
            "use_v2_block_manager": "",
            "disable_custom_all_reduce": ""
        },
        "client_parameters": {
            "model": "meta-llama/Meta-Llama-3-8B-Instruct",
            "backend": "vllm",
            "dataset_name": "sharegpt",
            "dataset_path": "./ShareGPT_V3_unfiltered_cleaned_split.json",
            "num_prompts": 100 
        }
    }
]

```

Reply:
One difference is that the EAGLE official implementation for some reason has dropped the input_layernorm for 1st decoder layer. Since in decoder layers we have this is not the case, it might be the reason for different results.

Reply:
maybe. whatever the issue is, it causes a big accuracy drop after k=1.